### PR TITLE
Convert to type script

### PR DESCRIPTION
**Pull request by russmiles**

### CODE REVIEW

File: .atomist/manifest.yml [hunk 1]
```diff
@@ -1,6 +1,6 @@
 group: atomist-rugs
 artifact: spring-boot-rest-service
-version: "0.7.1"
+version: "0.7.2"
```
Comment (ddgenome) [inline .atomist/manifest.yml]: Probably should be 0.8.0

File: .atomist.yml [hunk 2]
```diff
@@ -59,3 +59,32 @@ executor:
...
+  name: "AddTypeScriptGenerator"
+  group: "atomist-rugs"
+  artifact: "rug-editors"
+  version: "0.10.0"
+  origin:
+    repo: "https://github.com/atomist-rugs/rug-editors.git"
+    branch: "1979f9f98cad0eaf81eec3f512a80fdf3ae7d47f"
+    sha: "1979f9f"
+  parameters:
+    - "generator_name": "NewSpringBootRestServiceProj"
```
Comment (ddgenome) [inline .atomist.yml:88]: Was this editor run with the wrong generator name?

File: .atomist/editors/NewSpringBootRestService.ts [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+import { PopulateProject } from '@atomist/rug/operations/ProjectGenerator'
+import { Project } from '@atomist/rug/model/Core'
+import { Pattern } from '@atomist/rug/operations/RugOperation'
+import { Generator, Parameter, Tags } from '@atomist/rug/operations/Decorators'
+import { PathExpression, PathExpressionEngine } from '@atomist/rug/tree/PathExpression'
+import { File } from '@atomist/rug/model/File'
+
+@Generator("NewSpringBootRestService", "create a new Spring Boot Rest Service project")
+@Tags("documentation", "java", "spring", "spring-boot", "spring-rest")
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: Should probably remove the "documentation" tag.

File: .atomist/editors/NewSpringBootRestService.ts [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        validInput: "a valid GitHub project name consisting of alphanumeric, ., -, and _ characters",
+        minLength: 1,
+        maxLength: 100
+    })
+    project_name: string;
+
+    @Parameter({
+        displayName: "Maven Artifact ID",
+        description: "Maven artifact identifier, i.e., the name of the jar without the version, it is often the same as the project name",
+        pattern: "^[a-z][-a-z0-9_]*$", // Ideally this should be looking up artifact_id as a common pattern
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: Isn't there a set pattern for this? Oh, perhaps this was the one that was missing that you created an issue for?

File: .atomist/editors/NewSpringBootRestService.ts [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        minLength: 1,
+        maxLength: 21,
+        required: false
+    })
+    artifact_id: string = "myartifact";
+
+    @Parameter({
+        displayName: "Group ID",
+        description: "Maven group identifier, often used to provide a namespace for your project, e.g., com.pany.team",
+        pattern: "^[a-z][-a-z0-9_]*$", // Ideally this should be looking up artifact_id as a common pattern
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: I'm pretty sure there is a `Pattern.group_id` regex.

File: .atomist/editors/NewSpringBootRestService.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+    })
+    group_id: string = "mygroup";
+
+    @Parameter({
+        displayName: "Version",
+        description: "initial version of the project, e.g., 1.2.3-SNAPSHOT",
+        pattern: Pattern.semantic_version,
+        validInput: "a valid semantic version, http://semver.org",
+        minLength: 1,
+        maxLength: 21,
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: Is 21 long enough?

File: .atomist/editors/NewSpringBootRestService.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    })
+    version: string = "0.1.0-SNAPSHOT";
+
+    @Parameter({
+        displayName: "Project Description",
+        description: "short descriptive text describing the new project",
+        pattern: Pattern.any,
+        validInput: "free text",
+        minLength: 1,
+        maxLength: 50,
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: Typically we limit this to 100

File: .atomist/editors/NewSpringBootRestService.ts [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+    })
+    description: string = "My new project"
+
+    @Parameter({
+        displayName: "Root Package",
+        description: "root package for your generated source, often this will be namespaced under the group ID",
+        pattern: Pattern.java_package,
+        validInput: "a valid Java package name, which consists of period-separated identifiers which have only alphanumeric characters, $ and _ and do not start with a number",
+        minLength: 1,
+        maxLength: 50,
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts:78]: Is 50 long enough?

File: .atomist/editors/NewSpringBootRestService.ts [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+        this.renameServiceClass(project, this.service_class_name)
+
+        this.addCreationDateToReadme(project)
+
+        this.tidyProject(project);
+    }
+
+    switchREADME(project) {
+        project.merge("readme.vm", "README.md", {"project_name": this.project_name, "description": this.description});
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: Perhaps a good time to move away from templates too?

File: .atomist/editors/NewSpringBootRestService.ts [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        this.packageMove(project, this.root_package);
+
+        this.renameServiceClass(project, this.service_class_name)
+
+        this.addCreationDateToReadme(project)
+
+        this.tidyProject(project);
+    }
+
+    switchREADME(project) {
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: Shouldn't all of these function signatures provide a type for `project`, i.e., `project: Project`?

File: .atomist/editors/NewSpringBootRestService.ts [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        project.editWith("atomist-rugs.common-editors.PomParameterizer", {
+            "artifact_id": this.artifact_id, 
+            "group_id": this.group_id,
+            "version": this.version,
+            "name": this.project_name,
+            "description": this.description
+        });
+    }
+
+    packageMove(project, new_package) {
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: `new_project` is a `string`? TypeScript standards requires camelCase for function and variable names. So `new_package` should be `newPackage`.

Similar for all variables below.

**Review by ddgenome — CHANGES_REQUESTED** [review]
Shouldn't the `NewSpringBootRestService.rug` file have been removed rather than just moved to the top of the project?

File: .atomist/editors/NewSpringBootRestService.ts (hunk 4 above)
Comment (russmiles) [reply .atomist/editors/NewSpringBootRestService.ts]: Is indeed the one that was missing, for now.

File: .atomist/editors/NewSpringBootRestService.ts (hunk 6 above)
Comment (russmiles) [reply .atomist/editors/NewSpringBootRestService.ts]: Possibly not, lengthening to 50

File: .atomist/editors/NewSpringBootRestService.ts [hunk 12]
```diff
@@ -131,33 +140,37 @@ class NewSpringBootRestService implements PopulateProject {
...
-        let readmePathExpression = new PathExpression<Project, File>(
-            "/*[@name='README.md']");
+    addCreationDateToReadme(project: Project, readme: File) {
+        readme.replace("{{project_name}}", this.project_name);
 
-        let readmeFile: File = eng.scalar(project, readmePathExpression);
+        readme.replace("{{description}}", this.description);
 
-        readmeFile.replace("{{creation_date}}", new Date().toISOString().split('T')[0]);
+        readme.replace("{{creation_date}}", new Date().toISOString().split('T')[0]);
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: If you're going to do it this way, you might as well keep using the template.

Here's an example that uses the README from the generator's archive, replacing and excising as appropriate.

**Review by ddgenome — COMMENTED** [review]
Truly do not use templates.

File: .atomist/editors/NewSpringBootRestService.ts [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        this.switchREADME(project);
+
+        let readmeFile = this.selectReadmeFile(project)
+
+        this.addCreationDateToReadme(project, readmeFile)
+
+        this.tidyProject(project);
+    }
+
+    selectReadmeFile(project:Project) {
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: Shouldn't this also specify a return type?

```typescript
selectReadmeFile(project: Project): File {
```

File: .atomist/editors/NewSpringBootRestService.ts [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+        project.copyEditorBackingFileOrFail(defaultREADMEPath, readmeFileName);
+    }
+
+    addCreationDateToReadme(project: Project, readme: File) {
+        readme.replace("{{project_name}}", this.project_name);
+
+        readme.replace("{{description}}", this.description);
+
+        readme.replace("{{creation_date}}", new Date().toISOString().split('T')[0]);
+    }
```
Comment (ddgenome) [inline .atomist/editors/NewSpringBootRestService.ts]: If you're going to handle the README this way, you might as well keep using the template.

Here's an example that uses the README from the generator's archive, replacing and excising as appropriate.

**Review by ddgenome — COMMENTED** [review]
A few more questions

File: .atomist/editors/NewSpringBootRestService.ts (hunk 14 above)
Comment (russmiles) [reply .atomist/editors/NewSpringBootRestService.ts]: See your point, but don't see the example?

I'll move to pure README amendments instead, it makes sense.

Comment (ddgenome) [reply .atomist/editors/NewSpringBootRestService.ts]: Sorry copy/paste error. Here's the link to the example:
https://github.com/atomist-rugs/rug-project/blob/master/.atomist/editors/RugGeneratorFunctions.ts#L34-L44

Comment (russmiles) [reply .atomist/editors/NewSpringBootRestService.ts]: Thanks! I'll incorporate that approach right now.

File: README.md [hunk 15]
```diff
@@ -3,12 +3,7 @@
 [![Build Status](https://travis-ci.org/atomist-rugs/spring-boot-rest-service.svg?branch=master)](https://travis-ci.org/atomist-rugs/spring-boot-rest-service)
 [![Slack Status](https://join.atomist.com/badge.svg)](https://join.atomist.com)
 
-This [Rug][rug] archive contains a Generator for a [Spring Boot][boot]
-[Rest][rest] Service project.
-
-[rug]: http://docs.atomist.com/
-[boot]: https://projects.spring.io/spring-boot/
-[rest]: https://en.wikipedia.org/wiki/Representational_state_transfer
+This Rug archive contains a Generator for a Spring Boot Rest Service project.
```
Comment (ddgenome) [inline README.md]: We prefer lowercase generator.

File: .atomist/manifest.yml [hunk 16]
```diff
@@ -1,6 +1,6 @@
 group: atomist-rugs
 artifact: spring-boot-rest-service
-version: "0.7.1"
+version: "0.8.0"
 
 requires: "[0.10.0,1.0.0)"
```
Comment (ddgenome) [inline .atomist/manifest.yml]: Update to 0.12.0?

**Review by ddgenome — COMMENTED** [review]
Perhaps I should have made those functions editors so you could have called them instead of copying them.

A couple minor things and then it is good to merge.

**Comment by russmiles** [comment]
Like the idea of making those functions editors in the future. For now just those functions locally and if we decide that's the way to go it can be a new issue here and in the common-editors @ddgenome ?

**Comment by ddgenome** [comment]
Why did you remove the links to Spring Boot etc. in the README?

Sorry to be a pest but can you add an entry for these changes into the CHANGELOG and squash the commits?

**Comment by russmiles** [comment]
The links were removed as they made the regex heavily confused for replacement. Happy to reintroduce if I can get your help on making the regex replace work?

Good point on the CHANGELOG for sure, I'll get that entry done now and push.

**Comment by russmiles** [comment]
Commits will be squashed on "Squash and Merge"

File: .atomist/editors/RugGeneratorFunctions.ts [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+import { File } from '@atomist/rug/model/File'
+import { PathExpression, PathExpressionEngine } from '@atomist/rug/tree/PathExpression'
+
+export function cleanReadMe(project: Project, projectName: string, description: string, owner: string): void {
+    let eng: PathExpressionEngine = project.context().pathExpressionEngine();
+
+    let readMePE = new PathExpression<Project, File>("/*[@name='README.md']");
+    let readMe: File = eng.scalar(project, readMePE);
+    readMe.replace("# Atomist 'spring-boot-rest-service'", "# " + projectName);
+    readMe.regexpReplace("This Rug archive contains a generator for a Spring Boot Rest Service project.[\\s\\S]*?\n## Rugs\n", description + "\n");
```
Comment (ddgenome) [inline .atomist/editors/RugGeneratorFunctions.ts]: If you have the following text in the README:

```markdown
This [Rug][rug] project contains a generator for a [Spring Boot][boot]
[REST][rest] service project.

[rug]: http://docs.atomist.com/
[boot]: https://projects.spring.io/spring-boot/
[rest]: https://en.wikipedia.org/wiki/Representational_state_transfer

## Rugs
```

just alter the regex to skip the bad parts. Something like,

```typescript
readMe.regexpReplace("This .*?Rug.*? project contains a generator for a .*?Spring Boot[\\s\\S]*?\n## Rugs\n", description + "\n\n##Rugs\n");
```

**Review by ddgenome — COMMENTED** [review]
I added a comment on the text and regex.

File: .atomist/editors/RugGeneratorFunctions.ts (hunk 17 above)
Comment (russmiles) [reply .atomist/editors/RugGeneratorFunctions.ts]: Nice! Thanks @ddgenome ; I'll add those pieces in now.

File: .atomist/editors/RugGeneratorFunctions.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+import { File } from '@atomist/rug/model/File'
+import { PathExpression, PathExpressionEngine } from '@atomist/rug/tree/PathExpression'
+
+export function cleanReadMe(project: Project, projectName: string, description: string, owner: string): void {
+    let eng: PathExpressionEngine = project.context().pathExpressionEngine();
+
+    let readMePE = new PathExpression<Project, File>("/*[@name='README.md']");
+    let readMe: File = eng.scalar(project, readMePE);
+    readMe.replace("# Atomist 'spring-boot-rest-service'", "# " + projectName);
+    readMe.regexpReplace("This .*?Rug.*? project contains a generator for a .*?Spring Boot[\\s\\S]*?\n## Rugs\n", description + "\n");
```
Comment (ddgenome) [inline .atomist/editors/RugGeneratorFunctions.ts]: Seems you missed part of the replacement text. Shouldn't it be `"\n\n##Rugs\n"`?

**Review by ddgenome — COMMENTED** [review]
One more thing

File: .atomist/editors/RugGeneratorFunctions.ts (hunk 18 above)
Comment (russmiles) [reply .atomist/editors/RugGeneratorFunctions.ts]: I deliberately removed that as it produces an odd `## Rugs` header in the resulting README with no content as the generated project is not a rug archive.

Comment (ddgenome) [reply .atomist/editors/RugGeneratorFunctions.ts]: Yes, of course. We should probably simplify the code a bit then, eh? We really only need one replaceRegexp.

File: .atomist/editors/RugGeneratorFunctions.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    readMe.regexpReplace("\n### NewSpringBootRestService[\\s\\S]*\n## Support\n", "\n## Support\n");
```
Comment (ddgenome) [inline .atomist/editors/RugGeneratorFunctions.ts]: The above two `regexpReplace` can be replaced with a single:

```typescript
readMe.regexpReplace("This .*?Rug.*? project contains a generator for a .*?Spring Boot[\\s\\S]*?\n## Support\n", description + "\n\n## Support\n");
```

File: CHANGELOG.md [hunk 20]
```diff
@@ -7,7 +7,17 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
...
-[Unreleased]: https://github.com/atomist-rugs/spring-boot-rest-service/compare/0.6.0...HEAD
+[Unreleased]: https://github.com/atomist-rugs/spring-boot-rest-service/compare/0.7.1...HEAD
+
+Changes include:
+
+- Migrated generator to TypeScript and Rug 0.12.0.
+
+## [0.7.1] - 2017-01-03
+
+[0.7.0]: https://github.com/atomist-rugs/spring-boot-rest-service/compare/0.6.0...0.7.1
```
Comment (ddgenome) [inline CHANGELOG.md]: What version are we releasing? These numbers are all over the place. Is 0.6.0 that last release? If so, this should be 0.7.0.

**Review by ddgenome — COMMENTED** [review]
Combine statements and question about release version.

File: CHANGELOG.md (hunk 20 above)
Comment (russmiles) [reply CHANGELOG.md]: 0.7.1 is our last release that I can see?

**Review by ddgenome — APPROVED** [review]

**Pull request merged**